### PR TITLE
Closes #17 - Added global exception handling

### LINKED ISSUE #17: Add global exception handling

### PR DESCRIPTION
**Pull request by YurasNazar**
Added a global exception handling by creating "ErrorHandlingMiddleware.cs" and adding this midllware in Startup.cs . Also added folder extentions - in this folder created "Exceptons" folder , where other programmer can create and add own exception class. And Finally has been written unit tests by using NUnit.

### CODE REVIEW

File: gtd-timer/Extentions/ErrorHandlingMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            var result = JsonConvert.SerializeObject(new
+            {
+                error = exception.Message,
+                code  = exceptionCode
+            });
+            context.Response.ContentType = "application/json";
+            context.Response.StatusCode = (int)exceptionCode;
+            return context.Response.WriteAsync(result);
+        }
+        public HttpStatusCode GetExceptionCode(Exception exception)
```
Comment (roman-fay) [inline gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: this should be private

File: gtd-timer/Extentions/Exceptions/UserNotFoundException.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Runtime.Serialization;
+using System.Threading.Tasks;
+
+namespace gtdtimer.Extentions.Exceptions
+{
+    public class UserNotFoundException : Exception
```
[the remark this answers is not in the document]
Comment (YurasNazar) [reply gtd-timer/Extentions/Exceptions/UserNotFoundException.cs]: UserNotFoundException its my custom class and he is appointed to method Get , when we try to get some user by id , and such user exist then throw UserNotFoundException.@nataliapachva has UserExistException and uses this class when use the method Post , if such user already exist she will throw  UserExistException. So I do not think  we duplicate each other.

File: gtd-timer/Extentions/ErrorHandlingMiddleware.cs (hunk 1 above)
Comment (YurasNazar) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: But then I will not be able to test this method.

File: gtd-timer/Extentions/ErrorHandlingMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            HttpStatusCode exceptionCode;
+            string exceptionClassName = exception.GetType().ToString().Substring(exception.GetType().ToString().LastIndexOf('.') + 1);
```
[the remark this answers is not in the document]
Comment (YurasNazar) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: screenshot_1

[the remark this answers is not in the document]
Comment (YurasNazar) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: Error	CS0119	'UserNotFoundException' is a type, which is not valid in the given context	gt

[the remark this answers is not in the document]
Comment (roman-fay) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: Damn.. I typed it wrong. Try like this:

case UserNotFoundException _:

File: gtd-timer/Extentions/Exceptions/UserNotFoundException.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (roman-fay) [reply gtd-timer/Extentions/Exceptions/UserNotFoundException.cs]: agree

File: gtd-timer/Extentions/ErrorHandlingMiddleware.cs (hunk 1 above)
Comment (roman-fay) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: You would.. through HandleExceptionAsync. Look into using DefaultHttpContext as fake context for test.
I think something like this should be possible:

var context = new DefaultHttpContext();
context.Response.Body = new MemoryStream()
await errorHandlingMiddleware.Invoke(context)

and then reading context.Response.Body

Comment (roman-fay) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: You might also need to mock ErrorHandlingMiddleware...

File: gtd-timer/Extentions/ErrorHandlingMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+using Newtonsoft.Json;
+using gtdtimer.Extentions.Exceptions;
+
+namespace gtdtimer.Extentions
+{
+    public class ErrorHandlingMiddleware
+    {
+        private readonly RequestDelegate next;
+
+        public ErrorHandlingMiddleware(){}
```
Comment (roman-fay) [inline gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: btw, is this for unit test only?

Comment (YurasNazar) [reply gtd-timer/Extentions/ErrorHandlingMiddleware.cs]: Yes , I understand that is'nt the best way , So I will fix it.

**Pull request closed without merging**